### PR TITLE
fix: core_runner: Do not compete with NamedTemporaryFile objects for temp file removal.

### PR DESCRIPTION
**Pull request by emcd**
re: #5996 (cc: @emjin)

What this fix addresses is described in the changelog entry.
tldr; References to `NamedTemporaryFile` objects were not being maintained while processing occurred with their corresponding temp files. This allowed them to be garbage-collected during processing, which caused the temp files to disappear while workers or the core runner were trying to use them. And bad things would happen as a result.

Tested manually and fix seems good. (20 iterations with PyPy and not a single worker crash. Used updated code in an existing installation.)

Also built Docker image and ran a container from it:
```
$ git submodule update --init
$ make build-docker
$ docker run -it -v $(pwd):/src semgrep /bin/bash
```
Inside container, installed Pytest and some other dependencies and then ran Pytest:
```
$ apk add make
$ pip install pytest-snapshot pytest-xdist pytest-mock pytest-freezegun pytest-split pytest-rerunfailures
$ pytest -n auto --snapshot-update --allow-snapshot-deletion tests
$ pytest
...
============================================================================================== short test summary info ==============================================================================================
FAILED tests/e2e/test_check.py::test_hidden_rule__implicit - AssertionError: value does not match the expected value in snapshot /src/cli/tests/e2e/snapshots/test_check/test_hidden_rule__implicit/error.txt
FAILED tests/e2e/test_ci.py::test_full_run[autofix-github-pr] - AssertionError: value does not match the expected value in snapshot /src/cli/tests/e2e/snapshots/test_ci/test_full_run/autofix-github-pr/complete....
FAILED tests/e2e/test_ci.py::test_full_run[autofix-gitlab] - AssertionError: value does not match the expected value in snapshot /src/cli/tests/e2e/snapshots/test_ci/test_full_run/autofix-gitlab/complete.json
FAILED tests/e2e/test_ci.py::test_full_run[noautofix-azure-pipelines] - AssertionError: value does not match the expected value in snapshot /src/cli/tests/e2e/snapshots/test_ci/test_full_run/noautofix-azure-pip...
============================================================================= 4 failed, 1145 passed, 2322 warnings in 220.23s (0:03:40) =============================================================================
```
Mostly passed. Failures don't seem to be related to anything under my control or related to my changes.

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by aryx** [comment]
This looks great! @emjin what do you think?
Thx a lot for fixing those tricky bugs!

**Comment by emjin** [comment]
Thanks so much for investigating and creating a fix for this! I'll look over it in more detail asap on Monday

**Review by emjin — APPROVED** [review]
This is great! Thank you so much for your contribution!

**Comment by emjin** [comment]
@emcd If you'd like some swag, email me a way we can dm you a link! You can message me at emma@r2c.dev
